File: scripts/geometric_table_extractor.py
#!/usr/bin/env python3
import json
import sys
from typing import List, Dict

AnalysisHeaders = ["ANALIZA", "TEST", "DENUMIRE", "ANALIZE"]
ResultHeaders = ["REZULTAT", "VALOARE"]
UnitHeaders = ["UNITATE", "U.M.", "UM"]
ReferenceHeaders = ["REFERINTA", "REFERINȚĂ", "REF", "VALORI"]
RowToleranceY = 0.012


def cluster_into_rows(tokens: List[Dict]) -> List[List[Dict]]:
    sorted_tokens = sorted(tokens, key=lambda t: (t['boundingBox']['y'] + t['boundingBox']['height']/2.0))
    rows: List[List[Dict]] = []
    current_row = None
    current_row_y = None
    for t in sorted_tokens:
        center_y = t['boundingBox']['y'] + t['boundingBox']['height']/2.0
        if current_row is None or abs(center_y - current_row_y) > RowToleranceY:
            current_row = []
            rows.append(current_row)
            current_row_y = center_y
        current_row.append(t)
    for row in rows:
        row.sort(key=lambda a: a['boundingBox']['x'])
    return rows


def find_header_row(rows: List[List[Dict]]) -> int:
    for i, row in enumerate(rows):
        row_text = " ".join([t['text'] for t in row]).upper()
        has_analysis = any(h in row_text for h in AnalysisHeaders)
        has_result = any(h in row_text for h in ResultHeaders)
        if has_analysis and has_result:
            return i
    return -1


def determine_columns(header_row: List[Dict]):
    analysis_range = result_range = unit_range = reference_range = None
    for token in header_row:
        upper = token['text'].upper()
        bb = token['boundingBox']
        start = int(bb['x'] * 10000)
        end = int((bb['x'] + bb['width']) * 10000)
        if any(h in upper for h in AnalysisHeaders):
            analysis_range = (start, end)
        elif any(h in upper for h in ResultHeaders):
            result_range = (start, end)
        elif any(h in upper for h in UnitHeaders):
            unit_range = (start, end)
        elif any(h in upper for h in ReferenceHeaders):
            reference_range = (start, end)
    return analysis_range, result_range, unit_range, reference_range


def parse_data_row(row: List[Dict], columns):
    analysis_name = result_value = unit = reference = None
    for token in row:
        center_x = int((token['boundingBox']['x'] + token['boundingBox']['width']/2.0) * 10000)
        if columns[0] and columns[0][0] <= center_x <= columns[0][1]:
            analysis_name = (analysis_name + ' ' + token['text']) if analysis_name else token['text']
        elif columns[1] and columns[1][0] <= center_x <= columns[1][1]:
            result_value = (result_value + ' ' + token['text']) if result_value else token['text']
        elif columns[2] and columns[2][0] <= center_x <= columns[2][1]:
            unit = (unit + ' ' + token['text']) if unit else token['text']
        elif columns[3] and columns[3][0] <= center_x <= columns[3][1]:
            reference = (reference + ' ' + token['text']) if reference else token['text']
    if not analysis_name or not result_value:
        return None
    return {'analysis': analysis_name.strip(), 'value': result_value.strip(), 'unit': unit.strip() if unit else None, 'reference': reference.strip() if reference else None}


def extract(graph: Dict):
    pages = graph.get('pages', [])
    # use first page's tokens
    tokens = graph.get('allTokens', [])
    if not tokens:
        return []
    rows = cluster_into_rows(tokens)
    header_idx = find_header_row(rows)
    if header_idx < 0:
        return []
    columns = determine_columns(rows[header_idx])
    if columns[0] is None or columns[1] is None:
        return []
    lab_results = []
    for i in range(header_idx+1, len(rows)):
        r = parse_data_row(rows[i], columns)
        if r:
            lab_results.append(r)
    return lab_results


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Usage: geometric_table_extractor.py <native_debug_in.json>')
        sys.exit(1)
    inpath = sys.argv[1]
    j = json.load(open(inpath, 'r', encoding='utf-8'))
    graph = j.get('graph')
    if not graph:
        print('No graph found in input JSON')
        sys.exit(2)
    out = extract(graph)
    print(json.dumps(out, ensure_ascii=False, indent=2))

File: Mobile/DigitalTwin.Mobile.OCR/Services/GeometricTableExtractor.cs
using System.Globalization;
using System.Text.RegularExpressions;
using DigitalTwin.Mobile.OCR.Models.Graph;
using DigitalTwin.Mobile.OCR.Models.Structured;

namespace DigitalTwin.Mobile.OCR.Services;

/// <summary>
/// Extracts lab result rows from an OcrDocumentGraph using geometric bounding-box alignment.
/// No ML required — clusters tokens by Y-coordinate into rows, identifies column boundaries
/// from Romanian header keywords, then maps data tokens to columns.
/// </summary>
public sealed class GeometricTableExtractor
{
    private static readonly string[] AnalysisHeaders = ["ANALIZA", "TEST", "DENUMIRE", "ANALIZE"];
    private static readonly string[] ResultHeaders = ["REZULTAT", "VALOARE"];
    private static readonly string[] UnitHeaders = ["UNITATE", "U.M.", "UM"];
    private static readonly string[] ReferenceHeaders = ["REFERINTA", "REFERINȚĂ", "REF", "VALORI"];

    private const float RowToleranceY = 0.012f;

    private static string CleanTokenText(string s)
    {
        if (string.IsNullOrWhiteSpace(s)) return s ?? string.Empty;
        // Remove obvious OCR artefacts
        s = s.Replace('|', ' ').Replace('\u2013', '-').Replace('\u2014', '-');
        // Remove stray non-printable/control chars
        s = Regex.Replace(s, "[\u0000-\u001F]+", " ");
        // Collapse multiple spaces
        s = Regex.Replace(s, "\\s+", " ").Trim();
        return s;
    }

    // Try to split combined strings like "13.4 g/dL 11.5 - 17.5" into value, unit, reference
    private static (string Value, string? Unit, string? Reference) NormalizeValueUnitReference(string? rawValue, string? rawUnit, string? rawReference)
    {
        rawValue = CleanTokenText(rawValue ?? string.Empty);
        rawUnit = CleanTokenText(rawUnit ?? string.Empty);
        rawReference = CleanTokenText(rawReference ?? string.Empty);

        string? value = null;
        string? unit = string.IsNullOrWhiteSpace(rawUnit) ? null : rawUnit;
        string? reference = string.IsNullOrWhiteSpace(rawReference) ? null : rawReference;

        // If rawValue contains a leading numeric token, capture it
        var m = Regex.Match(rawValue, "([+-]?[0-9]+(?:[\\.,][0-9]+)?)");
        if (m.Success)
        {
            value = m.Groups[1].Value.Replace(',', '.');
            // remainder after the matched number may contain unit/reference
            var rem = rawValue.Substring(m.Index + m.Length).Trim();
            if (!string.IsNullOrWhiteSpace(rem))
            {
                // if we don't already have unit, try to take letters portion as unit
                if (unit is null)
                {
                    // split rem into unit and possible reference range by finding a numeric range
                    var rangeMatch = Regex.Match(rem, "([0-9]{1,3}(?:[\\.,][0-9]+)?\\s*[-–]\\s*[0-9]{1,3}(?:[\\.,][0-9]+)?)");
                    if (rangeMatch.Success)
                    {
                        reference = rangeMatch.Groups[1].Value.Replace(',', '.');
                        unit = rem.Substring(0, rangeMatch.Index).Trim();
                    }
                    else
                    {
                        unit = rem.Trim();
                    }
                }
                else
                {
                    // unit already present; if remainder contains range and reference empty, set it
                    if (reference is null)
                    {
                        var rangeMatch = Regex.Match(rem, "([0-9]{1,3}(?:[\\.,][0-9]+)?\\s*[-–]\\s*[0-9]{1,3}(?:[\\.,][0-9]+)?)");
                        if (rangeMatch.Success)
                            reference = rangeMatch.Groups[1].Value.Replace(',', '.');
                    }
                }
            }
        }
        else if (!string.IsNullOrWhiteSpace(rawValue))
        {
            // Nothing numeric in value cell; maybe unit cell contains the value+range
            var rangeMatch = Regex.Match(rawUnit ?? string.Empty, "([0-9]+(?:[\\.,][0-9]+)?(?:\\s*[-–]\\s*[0-9]+(?:[\\.,][0-9]+)?)?)");
            if (rangeMatch.Success)
            {
                value = rangeMatch.Groups[1].Value.Replace(',', '.');
                // remove from unit
                unit = Regex.Replace(rawUnit ?? string.Empty, Regex.Escape(rangeMatch.Groups[1].Value), "").Trim();
            }
            else
            {
                value = rawValue.Trim();
            }
        }

        // If unit contains a numeric range, split it
        if (!string.IsNullOrWhiteSpace(unit) && reference is null)
        {
            var rangeMatch2 = Regex.Match(unit, "([0-9]{1,3}(?:[\\.,][0-9]+)?\\s*[-–]\\s*[0-9]{1,3}(?:[\\.,][0-9]+)?)");
            if (rangeMatch2.Success)
            {
                reference = rangeMatch2.Groups[1].Value.Replace(',', '.');
                unit = Regex.Replace(unit, Regex.Escape(rangeMatch2.Groups[1].Value), "").Trim();
            }
        }

        // Final cleanups
        if (!string.IsNullOrWhiteSpace(value)) value = value.Trim();
        if (!string.IsNullOrWhiteSpace(unit)) unit = unit.Trim();
        if (!string.IsNullOrWhiteSpace(reference)) reference = reference.Trim();

        return (Value: value ?? string.Empty, Unit: unit, Reference: reference);
    }

    public IReadOnlyList<ExtractedLabResult> Extract(OcrDocumentGraph graph)
    {
        var results = new List<ExtractedLabResult>();
        foreach (var page in graph.Pages)
            results.AddRange(ExtractFromPage(page));
        return results;
    }

    private static IReadOnlyList<ExtractedLabResult> ExtractFromPage(OcrGraphPage page)
    {
        if (page.Tokens.Count == 0) return [];

        var rows = ClusterIntoRows(page.Tokens);
        int headerRowIdx = FindHeaderRow(rows);
        if (headerRowIdx < 0) return [];

        var columns = DetermineColumns(rows[headerRowIdx]);
        if (columns.AnalysisRange is null || columns.ResultRange is null) return [];

        var labResults = new List<ExtractedLabResult>();
        for (int i = headerRowIdx + 1; i < rows.Count; i++)
        {
            var labResult = ParseDataRow(rows[i], columns);
            if (labResult is not null)
                labResults.Add(labResult);
        }
        return labResults;
    }

    private static List<List<OcrToken>> ClusterIntoRows(IReadOnlyList<OcrToken> tokens)
    {
        var sorted = tokens.OrderBy(t => t.BoundingBox.CenterY).ToList();
        var rows = new List<List<OcrToken>>();
        List<OcrToken>? currentRow = null;
        float currentRowY = float.MinValue;

        foreach (var token in sorted)
        {
            if (currentRow is null || Math.Abs(token.BoundingBox.CenterY - currentRowY) > RowToleranceY)
            {
                currentRow = [];
                rows.Add(currentRow);
                currentRowY = token.BoundingBox.CenterY;
            }
            currentRow.Add(token);
        }

        foreach (var row in rows)
            row.Sort((a, b) => a.BoundingBox.X.CompareTo(b.BoundingBox.X));

        return rows;
    }

    private static int FindHeaderRow(List<List<OcrToken>> rows)
    {
        for (int i = 0; i < rows.Count; i++)
        {
            var rowText = string.Join(" ", rows[i].Select(t => t.Text)).ToUpperInvariant();
            bool hasAnalysis = AnalysisHeaders.Any(h => rowText.Contains(h));
            bool hasResult = ResultHeaders.Any(h => rowText.Contains(h));
            if (hasAnalysis && hasResult)
                return i;
        }
        return -1;
    }

    private static (Range? AnalysisRange, Range? ResultRange, Range? UnitRange, Range? ReferenceRange)
        DetermineColumns(List<OcrToken> headerRow)
    {
        Range? analysisRange = null, resultRange = null, unitRange = null, referenceRange = null;

        foreach (var token in headerRow)
        {
            var upper = token.Text.ToUpperInvariant();
            var bb = token.BoundingBox;
            var r = new Range((int)(bb.X * 10000), (int)(bb.Right * 10000));

            if (AnalysisHeaders.Any(h => upper.Contains(h)))
                analysisRange = r;
            else if (ResultHeaders.Any(h => upper.Contains(h)))
                resultRange = r;
            else if (UnitHeaders.Any(h => upper.Contains(h)))
                unitRange = r;
            else if (ReferenceHeaders.Any(h => upper.Contains(h)))
                referenceRange = r;
        }

        return (analysisRange, resultRange, unitRange, referenceRange);
    }

    private static ExtractedLabResult? ParseDataRow(
        List<OcrToken> row,
        (Range? AnalysisRange, Range? ResultRange, Range? UnitRange, Range? ReferenceRange) columns)
    {
        string? analysisName = null, resultValue = null, unit = null, referenceRange = null;

        foreach (var token in row)
        {
            int xScaled = (int)(token.BoundingBox.CenterX * 10000);
            var txt = CleanTokenText(token.Text);
            if (columns.AnalysisRange is { } ar && xScaled >= ar.Start.Value && xScaled <= ar.End.Value)
                analysisName = (analysisName is null) ? txt : $"{analysisName} {txt}";
            else if (columns.ResultRange is { } rr && xScaled >= rr.Start.Value && xScaled <= rr.End.Value)
                resultValue = (resultValue is null) ? txt : $"{resultValue} {txt}";
            else if (columns.UnitRange is { } ur && xScaled >= ur.Start.Value && xScaled <= ur.End.Value)
                unit = (unit is null) ? txt : $"{unit} {txt}";
            else if (columns.ReferenceRange is { } refR && xScaled >= refR.Start.Value && xScaled <= refR.End.Value)
                referenceRange = (referenceRange is null) ? txt : $"{referenceRange} {txt}";
        }

        if (string.IsNullOrWhiteSpace(analysisName) || string.IsNullOrWhiteSpace(resultValue))
            return null;

        var normalized = NormalizeValueUnitReference(resultValue, unit, referenceRange);
        var valueFinal = normalized.Value;
        var unitFinal = normalized.Unit;
        var referenceFinal = normalized.Reference;

        bool outOfRange = IsOutOfRange(valueFinal, referenceFinal);

        return new ExtractedLabResult(
            AnalysisName: new ExtractedField<string>(analysisName.Trim(), 0.85f, ExtractionMethod.BoundingBoxAlignment),
            Value: new ExtractedField<string>(valueFinal.Trim(), 0.85f, ExtractionMethod.BoundingBoxAlignment),
            Unit: unitFinal is null ? null : new ExtractedField<string>(unitFinal.Trim(), 0.80f, ExtractionMethod.BoundingBoxAlignment),
            ReferenceRange: referenceFinal is null ? null : new ExtractedField<string>(referenceFinal.Trim(), 0.80f, ExtractionMethod.BoundingBoxAlignment),
            SampleDate: null,
            IsOutOfRange: outOfRange);
    }

    private static bool IsOutOfRange(string value, string? referenceRange)
    {
        if (referenceRange is null) return false;
        if (!float.TryParse(value.Replace(',', '.'),
                System.Globalization.NumberStyles.Float,
                System.Globalization.CultureInfo.InvariantCulture, out var numVal))
            return false;

        var sep = referenceRange.Contains('–') ? '–' : '-';
        var parts = referenceRange.Split(sep, 2);
        if (parts.Length != 2) return false;

        bool loOk = float.TryParse(parts[0].Trim().Replace(',', '.'),
            System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out var lo);
        bool hiOk = float.TryParse(parts[1].Trim().Replace(',', '.'),
            System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out var hi);

        return loOk && hiOk && (numVal < lo || numVal > hi);
    }
}
